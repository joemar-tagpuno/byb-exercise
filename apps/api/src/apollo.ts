import { ApolloServer, type ApolloServerOptions } from '@apollo/server'
import { resolvers as scalarResolvers, typeDefs as scalarTypeDefs } from 'graphql-scalars'
import type { Container } from 'inversify'

import { merge } from 'lodash'
import { AssetManagerResolvers } from './modules/asset-manager/AssetManagerResolvers'
import { AssetManagerTypeDefs } from './modules/asset-manager/AssetManagerTypeDefs'

export function createApolloServer(kernel: Container) {
  const Query = `
        type ListMetadata {
          count: Int!
        }
        type Query {
          _empty: String
        },
        type Mutation {
          _empty: String
        }
    `

  const resolvers = merge(kernel.get(AssetManagerResolvers).resolvers(), {})

  const apolloServerConfig: ApolloServerOptions<any> = {
    typeDefs: [Query, AssetManagerTypeDefs, ...scalarTypeDefs],
    resolvers: {
      ...scalarResolvers,
      ...resolvers,
    },
    introspection: true, // for development environment only
  }

  return new ApolloServer(apolloServerConfig)
}
