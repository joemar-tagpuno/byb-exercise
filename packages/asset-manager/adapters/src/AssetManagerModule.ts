import { AssetManagerIdentifiers, CustomerSecureFileDownload } from '@byb/asset-manager-core'
import type { Container } from 'inversify'
import type { Sequelize } from 'sequelize'

import { DownloadableFileLoaderGateway } from './gateways/file-download/DownloadableFileLoaderGateway'
import { InMemoryDownloadableFileRepository } from './repositories/file-download/InMemoryDownloadableFileRepository'

export interface AssetManagerModuleConfiguration {
  sequelizeInstance: Sequelize
}

export class AssetManagerModule {
  constructor(private readonly container: Container) {}

  configure(config: AssetManagerModuleConfiguration) {
    // adapters
    this.container
      .bind(AssetManagerIdentifiers.downloadableFileLoader)
      .toConstantValue(new DownloadableFileLoaderGateway())
    this.container
      .bind(AssetManagerIdentifiers.downloadableFileRepository)
      .toConstantValue(new InMemoryDownloadableFileRepository(config.sequelizeInstance))

    // use-cases
    this.container.bind(CustomerSecureFileDownload).toSelf()
  }
}
