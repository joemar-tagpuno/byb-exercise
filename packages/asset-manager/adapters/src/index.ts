export { AssetManagerModule } from './AssetManagerModule'
export { DownloadableFileLoaderGateway } from './gateways/file-download/DownloadableFileLoaderGateway'
export { InMemoryDownloadableFileRepository } from './repositories/file-download/InMemoryDownloadableFileRepository'
