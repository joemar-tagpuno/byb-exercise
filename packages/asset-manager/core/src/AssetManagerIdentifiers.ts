export class AssetManagerIdentifiers {
  static readonly downloadableFileLoader = Symbol.for('asset_manager_downloadableFileLoader')
  static readonly downloadableFileRepository = Symbol.for('asset_manager_downloadableFileRepository')
}
