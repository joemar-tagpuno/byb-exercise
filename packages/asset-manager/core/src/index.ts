export { AssetManagerIdentifiers } from './AssetManagerIdentifiers'

// domain
export { AssetManagerErrors } from './domain/AssetManagerErrors'
export { DownloadableFile } from './domain/file-download/DownloadableFile'
export { DownloadableFileLoader } from './domain/file-download/DownloadableFileLoader'
export { DownloadableFileRepository } from './domain/file-download/DownloadableFileRepository'

// use-cases
export { CustomerSecureFileDownloadRequest, CustomerSecureFileDownload } from './usecases/CustomerSecureFileDownload'
