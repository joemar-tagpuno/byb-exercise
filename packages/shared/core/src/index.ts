export { UserIdentity, type UserRole } from './entities/UserIdentity'
export { Usecase } from './models/Usecase'
export { DomainError } from './models/DomainError'
