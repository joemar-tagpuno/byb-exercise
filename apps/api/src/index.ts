import path from 'node:path'
import { expressMiddleware } from '@apollo/server/express4'
import * as bodyParser from 'body-parser'
import cors from 'cors'
import express, { type Request, type Response } from 'express'
import { Sequelize } from 'sequelize'
import { createApolloServer } from './apollo'
import { BybKernel } from './config/BybKernel'
import config from './config/env'

const sequelizeInstance = new Sequelize({
  dialect: 'sqlite',
  storage: path.resolve(__dirname, './database/database.sqlite'),
})

// Allows to apply middleware on graphql resolvers
// e.g check user session or authentication token validation
// forward the authentication user data to resolvers as a context
// for resolver access restrictions
const createContextMiddleware = ({ req, res }: { req: Request; res: Response }) => {
  return { req, res }
}

async function bootstrap() {
  // configure application kernel
  const kernel = new BybKernel(sequelizeInstance)
  kernel.configure()

  // configure graphql server
  const apolloServer = createApolloServer(kernel)
  await apolloServer.start()

  // create express.js app instance
  const app = express()
  const port = config.port || 3000

  // apply basic API middlewares
  app.use(bodyParser.json())
  app.use(bodyParser.text({ type: 'application/graphql' }))
  app.use(
    bodyParser.urlencoded({
      parameterLimit: 100000,
      limit: '50mb',
      extended: true,
    }),
  )
  app.use(cors())

  // mount graphql endpoint
  app.use(
    '/graphql',
    expressMiddleware(apolloServer, {
      // @ts-ignore
      context: createContextMiddleware,
    }),
  )

  app.get('/', (_req: Request, res: Response) => {
    res.send('Hello, TypeScript with Express!')
  })

  // startup http server
  app.listen(port, () => {
    console.log(`Server is running on port ${port} in ${config.env} mode.`)
  })
}

bootstrap()
