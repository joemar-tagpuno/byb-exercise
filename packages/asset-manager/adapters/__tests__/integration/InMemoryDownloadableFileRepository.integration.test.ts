import { AssetManagerErrors } from '@byb/asset-manager-core'
import { beforeEach, describe, expect, it } from 'vitest'
import { InMemoryDownloadableFileRepository } from '../../src'

describe('InMemoryDownloadableFileRepository', () => {
  let underTest: InMemoryDownloadableFileRepository

  beforeEach(() => {
    underTest = new InMemoryDownloadableFileRepository()
  })

  describe('Everything is ok', () => {
    it('should return a file object', async () => {
      // WHEN
      const result = await underTest.findById('44a22e81-0ff7-471a-90be-0b49895cd1ca')

      // THEN
      expect(result).toEqual({
        name: 'file-1.pdf',
        url: 'http://localhost:3000/downloadable-files/44a22e81-0ff7-471a-90be-0b49895cd1ca',
        size: 140,
        type: 'pdf',
        provider: 'http',
      })
      expect(result).toMatchSnapshot()
    })

    it('should throw a model not found error result if record cannot be found', async () => {
      // WHEN
      const result = underTest.findById('unknown-file-id')

      // THEN
      await expect(result).rejects.toThrowError(AssetManagerErrors.FileDownload.DownloadableFileModelNotFound)
    })
  })
})
