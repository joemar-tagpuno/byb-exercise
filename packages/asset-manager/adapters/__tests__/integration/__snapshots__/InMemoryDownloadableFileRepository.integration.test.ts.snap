// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InMemoryDownloadableFileRepository > Everything is ok > should return a file object 1`] = `
{
  "name": "file-name.pdf",
  "provider": "http",
  "size": 124,
  "type": "pdf",
  "url": "http://localhost:123456/file/file-id",
}
`;
